### PR TITLE
feat: create Change Password modal

### PR DESCRIPTION
**Pull request by johnkim-det**
## Description

Create a modal that allows the user to change their password.

## Test Plan

Screen Shot 2022-02-01 at 10 03 15 AM

- [ ] User settings modal should display a "Change Password" button.

Screen Shot 2022-02-01 at 10 02 46 AM

- [ ] Clicking this button should open the "Change Password" modal.
- [ ] Form validation should check
  - Old password match
  - New password length (8 characters)
  - New password contains uppercase character
  - New password contains lowercase character
  - "Confirm password" value matches "New password" value
- [ ] Submitting a validated form should update the password

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: webui/react/src/services/types.ts [hunk 1]
```diff
@@ -265,3 +265,8 @@ export interface GetJobsResponse extends Api.V1GetJobsResponse {
 export interface GetJobQStatsParams extends FetchOptions {
   resourcePools?: string[];
 }
+
+export interface SetUserPasswordParams {
+  password: string,
+  username: string
```
Comment (EmilyBonar) [inline webui/react/src/services/types.ts:271]: nit: should probably end these lines with semicolons instead of commas (though both are valid)

File: webui/react/src/hooks/useModal/useModalUserSettings.tsx [hunk 2]
```diff
@@ -24,6 +28,12 @@ const useModalUserSettings = (): ModalHooks => {
           <span className={css.label}>Username</span>
           <span className={css.value}>{username}</span>
         </div>
+        <div className={css.field}>
+          <span className={css.label}>Password</span>
+          <span className={css.value}>
+            <Button onClick={() => openChangePasswordModal()}>Change password</Button>
```
Comment (EmilyBonar) [inline webui/react/src/hooks/useModal/useModalUserSettings.tsx:34]: nit: you should be able to pass the onClick without the callback function, e.g. `onClick={openChangePasswordModal}`

File: webui/react/src/hooks/useModal/useModalChangePassword.tsx [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  const [ form ] = Form.useForm();
+
+  const getModalContent = () => {
+    const submitForm = async (): Promise<void> => {
+      await setUserPassword({
+        password: form.getFieldValue('newPassword'),
+        username,
+      });
+      modalClose();
+    };
```
Comment (EmilyBonar) [inline webui/react/src/hooks/useModal/useModalChangePassword.tsx:23]: a couple of things:
1. why is `submitForm` defined inside `getModalContent` instead of just being a top level `useCallback` function inside of `useModalChangePassword`?
2. `await setUserPassword` should have a try/catch around it to handle any errors

File: webui/react/src/hooks/useModal/useModalChangePassword.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+              lowercase.
+            </span>
+          </Form.Item>
+          <Form.Item>
+            <div className={css.buttons}>
+              <Button onClick={() => onCancel()}>Cancel</Button>
+              <Button htmlType="submit" type="primary">
+                Change Password
+              </Button>
+            </div>
```
Comment (EmilyBonar) [inline webui/react/src/hooks/useModal/useModalChangePassword.tsx:96]: question: why are you defining your own buttons instead of just using the ones that should come with the modal?
also, nit: that onClick can just be `onClick={onCancel}`

File: webui/react/src/hooks/useModal/useModalChangePassword.tsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  const modalOpen = () => {
+    openOrUpdate({
+      className: css.noFooter,
+      closable: true,
+      content: getModalContent(),
+      icon: null,
+      onCancel,
+      title: 'Change Password',
+    });
+  };
```
Comment (EmilyBonar) [inline webui/react/src/hooks/useModal/useModalChangePassword.tsx:117]: nit: these should both be useCallback functions unless there's a reason they can't be
question: I see that you're removing the footer here in CSS, but why, when you're just re-implementing it yourself?

File: webui/react/src/hooks/useModal/useModalChangePassword.tsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            label="Old Password"
+            name="oldPassword"
+            required
+            rules={[
+              {
+                message: 'Incorrect password',
+                validator: async (rule, value) => {
+                  return await login({
+                    password: value || '',
+                    username: username || '',
```
Comment (EmilyBonar) [inline webui/react/src/hooks/useModal/useModalChangePassword.tsx:38]: nit: above you're setting `username = auth.user?.username || 'Anonymous';`, so why do you need the `|| ''` here?

File: webui/react/src/hooks/useModal/useModalChangePassword.tsx [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                    : Promise.reject();
+                },
+              },
+            ]}>
+            <Input.Password />
+          </Form.Item>
+          <Form.Item>
+            <span>
+              Password must be at least 8 charactes and contain one uppercase,
+              lowercase.
```
Comment (EmilyBonar) [inline webui/react/src/hooks/useModal/useModalChangePassword.tsx:87]: nit: maybe reword this, potentially `Password must be at least 8 characters and contain at least one uppercase and one lowercase character.`

**Review by EmilyBonar — CHANGES_REQUESTED** [review]
A good start, but I think a bit more polish will make it even better. I'd also recommend asking @alexjdetermined about the design of the settings modal, since there's a lot of blank space on the right side that I think could be improved.

File: webui/react/src/hooks/useModal/useModalUserSettings.tsx (hunk 2 above)
Comment (johnkim-det) [reply webui/react/src/hooks/useModal/useModalUserSettings.tsx]: i think type check requires a callback but changing it to a named function wrapped in useCallback

File: webui/react/src/hooks/useModal/useModalChangePassword.module.scss [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  }
+}
+.buttons {
+  display: flex;
+  justify-content: flex-end;
+
+  button + button {
+    margin-left: 8px;
+  }
+}
```
Comment (EmilyBonar) [inline webui/react/src/hooks/useModal/useModalChangePassword.module.scss:13]: nit: this css isn't used anymore

File: webui/react/src/hooks/useModal/useModalChangePassword.tsx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+                    : Promise.reject();
+                },
+              },
+            ]}>
+            <Input.Password />
+          </Form.Item>
+          <Form.Item>
+            <span>
+              Password must be at least 8 characters
+              and contain at least one uppercase and one lowercase character.
```
Comment (EmilyBonar) [inline webui/react/src/hooks/useModal/useModalChangePassword.tsx:97]: fix: Need to add that passwords are required to contain a number

**Review by EmilyBonar — APPROVED** [review]
LGTM! Just a couple of small changes
